packages/composable-controller/src/ComposableController.ts

```diff
@@ -1,75 +1,111 @@
-import type { RestrictedControllerMessenger } from '@metamask/base-controller';
-import { BaseControllerV1 } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type {
+  ControllerGetStateAction,
+  ControllerStateChangeEvent,
+  RestrictedControllerMessenger,
+  BaseControllerV1,
+} from '@metamask/base-controller';
 
-/**
- * List of child controller instances
- *
+const controllerName = 'ComposableController';
+
+/*
  * This type encompasses controllers based up either BaseControllerV1 or
  * BaseController. The BaseController type can't be included directly
  * because the generic parameters it expects require knowing the exact state
  * shape, so instead we look for an object with the BaseController properties
  * that we use in the ComposableController (name and state).
  */
-export type ControllerList = (
+type ControllerInstance =
   | BaseControllerV1<any, any>
-  | { name: string; state: Record<string, unknown> }
-)[];
+  | { name: string; state: Record<string, unknown> };
 
-export type ComposableControllerRestrictedMessenger =
-  RestrictedControllerMessenger<'ComposableController', never, any, never, any>;
+/**
+ * List of child controller instances
+ */
+export type ControllerList = ControllerInstance[];
 
 /**
- * Controller that can be used to compose multiple controllers together.
+ * Determines if the given controller is an instance of BaseControllerV1
+ * @param controller - Controller instance to check
+ * @returns True if the controller is an instance of BaseControllerV1
  */
-export class ComposableController extends BaseControllerV1<never, any> {
-  private readonly controllers: ControllerList = [];
+function isBaseControllerV1(
+  controller: ControllerInstance,
+): controller is BaseControllerV1<any, any> {
+  return 'subscribe' in controller && controller.subscribe !== undefined;
+}
 
-  private readonly messagingSystem?: ComposableControllerRestrictedMessenger;
+export type ComposableControllerState = {
+  [name: string]: ControllerInstance['state'];
+};
 
-  /**
-   * Name of this controller used during composition
-   */
-  override name = 'ComposableController';
+export type ComposableControllerGetStateAction = ControllerGetStateAction<
+  `${typeof controllerName}`,
+  ComposableControllerState
+>;
+
+export type ComposableControllerStateChangeEvent = ControllerStateChangeEvent<
+  `${typeof controllerName}`,
+  ComposableControllerState
+>;
+
+export type ComposableControllerActions = ComposableControllerGetStateAction;
+
+export type ComposableControllerEvents = ComposableControllerStateChangeEvent;
+
+export type ComposableControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  ControllerGetStateAction<string, Record<string, unknown>>,
```

Comment:
I removed the `getState` action and `ComposableControllerActions` entirely since they're not being used anywhere: 1ddf33cd10ec33a4c0a375b4749b0d2bbe87946e.